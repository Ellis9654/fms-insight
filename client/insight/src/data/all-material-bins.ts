/* Copyright (c) 2018, John Lenz

All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

    * Redistributions of source code must retain the above copyright
      notice, this list of conditions and the following disclaimer.

    * Redistributions in binary form must reproduce the above
      copyright notice, this list of conditions and the following
      disclaimer in the documentation and/or other materials provided
      with the distribution.

    * Neither the name of John Lenz, Black Maple Software, SeedTactics,
      nor the names of other contributors may be used to endorse or
      promote products derived from this software without specific
      prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

import * as api from "./api";
import * as im from "immutable";

export type MaterialList = ReadonlyArray<Readonly<api.IInProcessMaterial>>;

export type AllMaterialBins = im.Map<string, MaterialList>;

export function selectAllMaterialIntoBins(curSt: Readonly<api.ICurrentStatus>): AllMaterialBins {
  const palLoc = im
    .Map(curSt.pallets)
    .map(st => " (" + st.currentPalletLocation.group + " #" + st.currentPalletLocation.num.toString() + ")");

  return im
    .Seq(curSt.material)
    .map(mat => {
      switch (mat.location.type) {
        case api.LocType.InQueue:
          return { region: mat.location.currentQueue || "Queue", mat };
        case api.LocType.OnPallet:
          let region = "Pallet";
          if (mat.location.pallet) {
            region = "Pallet " + mat.location.pallet.toString() + palLoc.get(mat.location.pallet) || "";
          }
          return { region, mat };
        case api.LocType.Free:
        default:
          switch (mat.action.type) {
            case api.ActionType.Loading:
              return { region: "Raw Material", mat };
            default:
              return { region: "Free Material", mat };
          }
      }
    })
    .groupBy(x => x.region)
    .map(group =>
      group
        .map(x => x.mat)
        .sortBy(mat => {
          switch (mat.location.type) {
            case api.LocType.OnPallet:
              return mat.location.face;
            case api.LocType.InQueue:
            case api.LocType.Free:
            default:
              return mat.location.queuePosition;
          }
        })
        .valueSeq()
        .toArray()
    )
    .toMap();
}
