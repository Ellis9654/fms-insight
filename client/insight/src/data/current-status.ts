/* Copyright (c) 2018, John Lenz

All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

    * Redistributions of source code must retain the above copyright
      notice, this list of conditions and the following disclaimer.

    * Redistributions in binary form must reproduce the above
      copyright notice, this list of conditions and the following
      disclaimer in the documentation and/or other materials provided
      with the distribution.

    * Neither the name of John Lenz, Black Maple Software, SeedTactics,
      nor the names of other contributors may be used to endorse or
      promote products derived from this software without specific
      prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */
import * as api from './api';
import { ConsumingPledge, PledgeStatus } from './pledge';

export interface State {
    readonly loading: boolean;
    readonly loading_error?: Error;
    readonly current_status: Readonly<api.ICurrentStatus>; // TODO: DeepReadonly
}

const initial: State = {
    loading: false,
    current_status: {
        jobs: {},
        pallets: {},
        material: [],
        alarms: [],
        queues: {}
    }
};

export enum ActionType {
    NewCurrentStatus = 'CurStatus_NewCurrentStatus',
    Other = 'Other',
}

// TODO: Pledge and DeepReadonly in typescript 2.8
export type Action =
  | {type: ActionType.NewCurrentStatus, pledge: ConsumingPledge<Readonly<api.ICurrentStatus>> }
  | {type: ActionType.Other}
  ;

export function loadCurrentStatus() {
    var client = new api.JobsClient();
    return {
        type: ActionType.NewCurrentStatus,
        pledge: client.currentStatus()
    };
}

export function reducer(s: State, a: Action): State {
    if (s === undefined) { return initial; }
    switch (a.type) {
        case ActionType.NewCurrentStatus:
            switch (a.pledge.status) {
                case PledgeStatus.Starting:
                    return {...s, loading: true, loading_error: undefined};
                case PledgeStatus.Completed:
                    return {...s,
                        loading: false,
                        current_status: a.pledge.result,
                    };
                case PledgeStatus.Error:
                    return {...s, loading_error: a.pledge.error};

                default: return s;
            }
        default: return s;
    }
}