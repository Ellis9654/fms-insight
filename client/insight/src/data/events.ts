/* Copyright (c) 2018, John Lenz

All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

    * Redistributions of source code must retain the above copyright
      notice, this list of conditions and the following disclaimer.

    * Redistributions in binary form must reproduce the above
      copyright notice, this list of conditions and the following
      disclaimer in the documentation and/or other materials provided
      with the distribution.

    * Neither the name of John Lenz, Black Maple Software, SeedTactics,
      nor the names of other contributors may be used to endorse or
      promote products derived from this software without specific
      prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */
import * as api from './api';
import { duration } from 'moment';
import { PledgeStatus, ConsumingPledge } from './pledge';
import { addDays, addMonths, startOfMonth } from 'date-fns';
import * as im from 'immutable'; // consider collectable.js at some point?

export interface StationInUse {
    readonly date: Date;
    readonly station: string;
    readonly hours: number;
}

export enum AnalysisPeriod {
    Last30Days = 'Last_30_Days',
    SpecificMonth = 'Specific_Month'
}

export interface State {
    readonly loading_events: boolean;
    readonly loading_error?: Error;

    // list of entries sorted by endUTC
    readonly last_30_days_of_events: im.List<Readonly<api.ILogEntry>>; // TODO: deep readonly in typescript 2.8

    // list of station use, sorted by date
    readonly last_week_of_hours: im.List<StationInUse>;

    readonly system_active_hours_per_week: number;

    readonly analysis_period: AnalysisPeriod;
    readonly analysis_period_month: Date;
    readonly loading_analysis_period: boolean;
    readonly analysis_period_month_events: im.List<Readonly<api.ILogEntry>>;
}

export const initial: State = {
    loading_events: false,
    last_30_days_of_events: im.List(),
    last_week_of_hours: im.List(),
    system_active_hours_per_week: 7 * 24,
    analysis_period: AnalysisPeriod.Last30Days,
    analysis_period_month: startOfMonth(new Date()),
    loading_analysis_period: false,
    analysis_period_month_events: im.List()
};

export enum ActionType {
    RequestLastWeek = 'Events_RequestLastWeek',
    SetSystemHours = 'Events_SetSystemHours',
    LoadAnalysisLast30Days = 'Events_LoadAnalysisLast30Days',
    LoadAnalysisSpecificMonth = 'Events_LoadAnalysisMonth',
    SetAnalysisMonth = 'Events_SetAnalysisMonth',
    Other = 'Other',
}

// TODO: use Plege when typescript 2.8 shows up
export type Action =
  | {type: ActionType.RequestLastWeek, now: Date, pledge: ConsumingPledge<ReadonlyArray<Readonly<api.ILogEntry>>>}
  | {type: ActionType.SetSystemHours, hours: number}
  | {type: ActionType.LoadAnalysisLast30Days }
  | {
      type: ActionType.LoadAnalysisSpecificMonth,
      month: Date,
      pledge: ConsumingPledge<ReadonlyArray<Readonly<api.ILogEntry>>>
    }
  | {type: ActionType.SetAnalysisMonth, month: Date }
  | {type: ActionType.Other}
  ;

export function requestLastMonth() /*: Action<ActionUse.CreatingAction> */ {
    var client = new api.LogClient();
    var now = new Date();
    var oneWeekAgo = addDays(now, -30);
    return {
        type: ActionType.RequestLastWeek,
        now: now,
        pledge: client.get(oneWeekAgo, now)
    };
}

export function analyzeLast30Days() {
    return {type: ActionType.LoadAnalysisLast30Days};
}

export function analyzeSpecificMonth(month: Date) {
    var client = new api.LogClient();
    var startOfNextMonth = addMonths(month, 1);
    return {
        type: ActionType.LoadAnalysisSpecificMonth,
        month: month,
        pledge: client.get(month, startOfNextMonth)
    };
}

export function setAnalysisMonth(month: Date) {
    return {
        type: ActionType.SetAnalysisMonth,
        month: month
    };
}

function stat_name(e: api.ILogEntry): string | null {
    switch (e.type) {
        case api.LogType.LoadUnloadCycle:
        case api.LogType.MachineCycle:
            return e.loc + ' #' + e.locnum;
        default:
            return null;
    }
}

function refresh_weekly_hours(now: Date, newEvts: ReadonlyArray<api.ILogEntry>, st: State): State {
    let hours = st.last_week_of_hours;
    const oneWeekAgo = addDays(now, -7);

    // check if no changes needed: no new events and nothing to filter out
    const minEntry = hours.first();
    if ((minEntry === undefined || minEntry.date >= oneWeekAgo) && newEvts.length === 0) {
        return st;
    }

    // new entries
    const newEntries =
        im.Seq(newEvts)
        .filter(e => {
            if (e.endUTC < oneWeekAgo) { return false; }
            if (e.startofcycle) { return false; }
            let name = stat_name(e);
            if (!name) { return false; }
            let activeHrs = duration(e.active).asHours();
            if (activeHrs <=  0) { return false; }
            return true;
        })
        .map(e => ({
            date: e.endUTC,
            station: stat_name(e) || '',
            // TODO: calculate estimated active time
            hours: duration(e.active).asHours()
        }));

    hours = hours.toSeq()
        .filter(e => e.date >= oneWeekAgo)
        .concat(newEntries)
        .sortBy(e => e.date)
        .toList();

    return {...st, last_week_of_hours: hours};
}

function refresh_events(now: Date, newEvts: ReadonlyArray<api.ILogEntry>, st: State): State {
    let evts = st.last_30_days_of_events;
    let thirtyDaysAgo = addDays(now, -30);

    // check if no changes needed: no new events and nothing to filter out
    const minEntry = evts.first();
    if ((minEntry === undefined || minEntry.endUTC >= thirtyDaysAgo) && newEvts.length === 0) {
        return st;
    }

    evts = evts.toSeq()
        .concat(newEvts)
        .filter(e => e.endUTC >= thirtyDaysAgo)
        .sortBy(e => e.endUTC)
        .toList();

    return {...st, last_30_days_of_events: evts};
}

export function reducer(s: State, a: Action): State {
    if (s === undefined) { return initial; }
    switch (a.type) {
        case ActionType.RequestLastWeek:
            switch (a.pledge.status) {
                case PledgeStatus.Starting:
                    return {...s, loading_events: true, loading_error: undefined};
                case PledgeStatus.Completed:
                    let s2 = refresh_weekly_hours(a.now, a.pledge.result, s);
                    let s3 = refresh_events(a.now, a.pledge.result, s2);
                    return {...s3, loading_events: false};
                case PledgeStatus.Error:
                    return {...s, loading_events: false, loading_error: a.pledge.error};
                default: return s;
            }

        case ActionType.SetSystemHours:
            return {...s, system_active_hours_per_week: a.hours};

        case ActionType.LoadAnalysisLast30Days:
            return {...s,
                analysis_period: AnalysisPeriod.Last30Days,
                analysis_period_month_events: im.List()
            };

        case ActionType.LoadAnalysisSpecificMonth:
            switch (a.pledge.status) {
                case PledgeStatus.Starting:
                    return {...s,
                        analysis_period: AnalysisPeriod.SpecificMonth,
                        analysis_period_month: a.month,
                        loading_error: undefined,
                        loading_analysis_period: true,
                        analysis_period_month_events: im.List()
                    };

                case PledgeStatus.Completed:
                    return {...s,
                        loading_analysis_period: false,
                        analysis_period_month_events:
                            im.List(a.pledge.result).sortBy(e => e.endUTC)
                    };

                case PledgeStatus.Error:
                    return {...s,
                        loading_analysis_period: false,
                        loading_error: a.pledge.error
                    };

                default: return s;
            }

        case ActionType.SetAnalysisMonth:
            return {...s, analysis_period_month: a.month};

        default: return s;
    }
}