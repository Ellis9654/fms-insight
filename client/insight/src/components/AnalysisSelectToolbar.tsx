/* Copyright (c) 2018, John Lenz

All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

    * Redistributions of source code must retain the above copyright
      notice, this list of conditions and the following disclaimer.

    * Redistributions in binary form must reproduce the above
      copyright notice, this list of conditions and the following
      disclaimer in the documentation and/or other materials provided
      with the distribution.

    * Neither the name of John Lenz, Black Maple Software, SeedTactics,
      nor the names of other contributors may be used to endorse or
      promote products derived from this software without specific
      prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */
import * as React from 'react';
import Radio from 'material-ui/Radio';
import TextField from 'material-ui/TextField';
import { FormControlLabel } from 'material-ui/Form';
import { connect } from 'react-redux';
import { startOfMonth, format, parse } from 'date-fns';

import * as events from '../data/events';
import { Store } from '../data/store';

const toolbarStyle = {
  'display': 'flex',
  'backgroundColor': '#E0E0E0',
  'paddingLeft': '24px',
  'paddingRight': '24px',
  'minHeight': '2.5em',
  'alignItems': 'center' as 'center',
  'justifyContent': 'space-evenly' as 'space-evenly'
};

export interface AnalysisSelectToolbarProps {
  period: events.AnalysisPeriod;
  period_month: Date;
  // tslint:disable-next-line:no-any
  analyzeLast30Days: () => any;
  // tslint:disable-next-line:no-any
  analyzeMonth: (month: Date) => any;
  // tslint:disable-next-line:no-any
  setMonth: (month: Date) => any;
}

export class AnalysisSelectToolbar extends React.PureComponent<AnalysisSelectToolbarProps, {temp_month?: Date}> {

  state = {temp_month: undefined};

  setTempMonth = (m: Date) => {
    this.setState({temp_month: startOfMonth(m)});
  }

  blurMonth = () => {
    var m = this.state.temp_month;
    if (m === undefined) { return; }
    if (this.props.period_month === m) { return; }

    if (this.props.period === events.AnalysisPeriod.SpecificMonth) {
      // if month type is selected, reload data
      this.props.analyzeMonth(m);
    } else {
      // otherwise, just store month
      this.props.setMonth(m);
    }

    this.setState({temp_month: undefined});
  }

  render() {
    const curMonth: Date = this.state.temp_month || this.props.period_month;

    return (
      <nav style={toolbarStyle}>
        <FormControlLabel
          control={
            <Radio
              checked={this.props.period === events.AnalysisPeriod.Last30Days}
              onChange={e => e.target.value ? this.props.analyzeLast30Days() : null}
            />
          }
          label="Last 30 days"
        />
        <div>
          <FormControlLabel
            control={
              <Radio
                checked={this.props.period === events.AnalysisPeriod.SpecificMonth}
                onChange={e => e.target.value ? this.props.analyzeMonth(curMonth) : null}
              />}
            label="Select Month"
          />
          <TextField
            type="month"
            value={format(curMonth, 'YYYY-MM')}
            onChange={m => this.setTempMonth(parse(m.target.value))}
            onBlur={() => this.blurMonth()}
          />
        </div>
      </nav>
    );
  }
}

export default connect(
  (s: Store) => ({
    period: s.Events.analysis_period,
    period_month: s.Events.analysis_period_month,
  }),
  {
    analyzeLast30Days: events.analyzeLast30Days,
    analyzeMonth: events.analyzeSpecificMonth,
    setMonth: events.setAnalysisMonth
  }
)(AnalysisSelectToolbar);