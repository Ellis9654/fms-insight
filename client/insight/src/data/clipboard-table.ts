/* Copyright (c) 2019, John Lenz

All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

    * Redistributions of source code must retain the above copyright
      notice, this list of conditions and the following disclaimer.

    * Redistributions in binary form must reproduce the above
      copyright notice, this list of conditions and the following
      disclaimer in the documentation and/or other materials provided
      with the distribution.

    * Neither the name of John Lenz, Black Maple Software, SeedTactics,
      nor the names of other contributors may be used to endorse or
      promote products derived from this software without specific
      prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

import { LazySeq } from "./lazyseq";
import { fieldsHashCode } from "prelude-ts";
import { FilteredStationCycles, stat_name_and_num } from "./events.cycles";
import { format } from "date-fns";
const copy = require("copy-to-clipboard");

export interface ClipboardTablePoint {
  readonly x: Date;
  readonly y: string;
  readonly label: string;
}

class PointsTableCell {
  public constructor(public readonly x: number, public readonly y: string) {}
  equals(other: PointsTableCell): boolean {
    return this.x === other.x && this.y === other.y;
  }
  hashCode(): number {
    return fieldsHashCode(this.x, this.y);
  }
  toString(): string {
    return `{x: ${new Date(this.x).toISOString()}, y: ${this.y}}`;
  }
}

export function buildPointsTable(yTitle: string, points: ReadonlyArray<ClipboardTablePoint>): string {
  const cells = LazySeq.ofIterable(points).toMap(
    p => [new PointsTableCell(p.x.getTime(), p.y), p],
    (_, c) => c // cells should be unique, but just in case take the second
  );
  const days = LazySeq.ofIterable(points)
    .toSet(p => p.x.getTime())
    .toArray({ sortOn: x => x });
  const rows = LazySeq.ofIterable(points)
    .toSet(p => p.y)
    .toArray({ sortOn: x => x });

  let table = "<table>\n<thead><tr><th>" + yTitle + "</th>";
  for (let x of days) {
    table += "<th>" + new Date(x).toDateString() + "</th>";
  }
  table += "</tr></thead>\n<tbody>\n";
  for (let y of rows) {
    table += "<tr><th>" + y + "</th>";
    for (let x of days) {
      const cell = cells.get(new PointsTableCell(x, y));
      if (cell.isSome()) {
        table += "<td>" + cell.get().label + "</td>";
      } else {
        table += "<td></td>";
      }
    }
    table += "</tr>\n";
  }
  table += "</tbody>\n</table>";
  return table;
}

export function copyPointsToClipboard(yTitle: string, points: ReadonlyArray<ClipboardTablePoint>): void {
  copy(buildPointsTable(yTitle, points));
}

export function buildCycleTable(cycles: FilteredStationCycles): string {
  let table = "<table>\n<thead><tr>";
  table += "<th>Date</th><th>Part</th><th>Station</th><th>Pallet</th>";
  table += "<th>Serial</th><th>Workorder</th><td>Elapsed Min</td><td>Active Min</td>";
  table += "</tr></thead>\n<tbody>\n";
  for (let cycle of LazySeq.ofIterable(cycles.data).flatMap(([_, c]) => c)) {
    table += "<tr>";
    table += "<td>" + format(cycle.x, "MMM D, YYYY, H:mm a") + "</td>";
    table += "<td>" + cycle.part + "-" + cycle.process.toString() + "</td>";
    table += "<td>" + stat_name_and_num(cycle.stationGroup, cycle.stationNumber) + "</td>";
    table += "<td>" + cycle.pallet + "</td>";
    table += "<td>" + (cycle.serial || "") + "</td>";
    table += "<td>" + (cycle.workorder || "") + "</td>";
    table += "<td>" + cycle.y.toFixed(1) + "</td>";
    table += "<td>" + cycle.active.toFixed(1) + "</td>";
    table += "</tr>\n";
  }
  table += "</tbody>\n</table>";
  return table;
}

export function copyCyclesToClipboard(cycles: FilteredStationCycles): void {
  copy(buildCycleTable(cycles));
}
