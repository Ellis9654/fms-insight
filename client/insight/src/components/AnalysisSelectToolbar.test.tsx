/* Copyright (c) 2018, John Lenz

All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

    * Redistributions of source code must retain the above copyright
      notice, this list of conditions and the following disclaimer.

    * Redistributions in binary form must reproduce the above
      copyright notice, this list of conditions and the following
      disclaimer in the documentation and/or other materials provided
      with the distribution.

    * Neither the name of John Lenz, Black Maple Software, SeedTactics,
      nor the names of other contributors may be used to endorse or
      promote products derived from this software without specific
      prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

import * as React from 'react';
import { shallow } from 'enzyme';

import { AnalysisSelectToolbar } from './AnalysisSelectToolbar';
import * as events from '../data/events';

it('displays the toolbar', () => {
  const analyzeLast30 = jest.fn();
  const analyzeMonth = jest.fn();
  const setMonth = jest.fn();

  let val = shallow(
    <AnalysisSelectToolbar
      period={events.AnalysisPeriod.Last30Days}
      period_month={new Date(2018, 0, 1)}
      analyzeLast30Days={analyzeLast30}
      analyzeMonth={analyzeMonth}
      setMonth={setMonth}
    />);
  expect(val).toMatchSnapshot('last 30 selected with january');

  val = shallow(
    <AnalysisSelectToolbar
      period={events.AnalysisPeriod.SpecificMonth}
      period_month={new Date(2018, 1, 1)}
      analyzeLast30Days={analyzeLast30}
      analyzeMonth={analyzeMonth}
      setMonth={setMonth}
    />);
  expect(val).toMatchSnapshot('specific month with february');
});

it('operates when period is last 30', () => {
  const month = new Date(2018, 1, 1);
  const analyzeLast30 = jest.fn();
  const analyzeMonth = jest.fn();
  const setMonth = jest.fn();

  let val = shallow(
    <AnalysisSelectToolbar
      period={events.AnalysisPeriod.Last30Days}
      period_month={month}
      analyzeLast30Days={analyzeLast30}
      analyzeMonth={analyzeMonth}
      setMonth={setMonth}
    />);

  // tslint:disable-next-line
  const textFieldChange = val.find('TextField').prop('onChange') as any;
  // tslint:disable-next-line
  const textFieldBlur = val.find('TextField').prop('onBlur') as any;

  // switch to month
  val
    .find({label: "Select Month"})
    .prop("control")
    .props.onChange(null, {target: {value: true}});
  expect(analyzeMonth).toHaveBeenCalledWith(month);
  analyzeMonth.mockReset();

  // change month should set but not analyze
  textFieldChange({target: {value: '2018-05'}});
  expect(setMonth).not.toHaveBeenCalled();
  expect(analyzeMonth).not.toHaveBeenCalled();

  textFieldBlur();
  expect(setMonth).toHaveBeenCalledWith(new Date(2018, 4, 1));
  expect(analyzeMonth).not.toHaveBeenCalled();
  setMonth.mockReset();
});

it("operates when period is specific month", () => {
  const month = new Date(2018, 2, 1);
  const analyzeLast30 = jest.fn();
  const analyzeMonth = jest.fn();
  const setMonth = jest.fn();

  let val = shallow(
    <AnalysisSelectToolbar
      period={events.AnalysisPeriod.SpecificMonth}
      period_month={month}
      analyzeLast30Days={analyzeLast30}
      analyzeMonth={analyzeMonth}
      setMonth={setMonth}
    />);

  // tslint:disable-next-line
  const textFieldChange = val.find('TextField').prop('onChange') as any;
  // tslint:disable-next-line
  const textFieldBlur = val.find('TextField').prop('onBlur') as any;

  // now set back to last 30 days
  val
    .find({label: "Last 30 days"})
    .prop("control")
    .props.onChange(null, {target: {value: true}});
  expect(analyzeLast30).toHaveBeenCalled();
  analyzeLast30.mockReset();

  // change month
  textFieldChange({target: {value: '2018-06'}});
  expect(setMonth).not.toHaveBeenCalled();
  expect(analyzeMonth).not.toHaveBeenCalled();

  // blur month, which should reload
  textFieldBlur();
  expect(setMonth).not.toHaveBeenCalled();
  expect(analyzeMonth).toHaveBeenCalledWith(new Date(2018, 5, 1));
  analyzeMonth.mockReset();
});