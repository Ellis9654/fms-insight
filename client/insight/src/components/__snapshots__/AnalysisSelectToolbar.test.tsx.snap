// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`last 30 selected with january 1`] = `
<nav
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#E0E0E0",
      "display": "flex",
      "justifyContent": "space-evenly",
      "minHeight": "2.5em",
      "paddingLeft": "24px",
      "paddingRight": "24px",
    }
  }
>
  <WithStyles(FormControlLabel)
    control={
      <WithStyles(Radio)
        checked={true}
        onChange={[Function]}
      />
    }
    label="Last 30 days"
  />
  <div>
    <WithStyles(FormControlLabel)
      control={
        <WithStyles(Radio)
          checked={false}
          onChange={[Function]}
        />
      }
      label="Select Month"
    />
    <TextField
      onBlur={[Function]}
      onChange={[Function]}
      required={false}
      select={false}
      type="month"
      value="2018-01"
    />
  </div>
</nav>
`;

exports[`specific month with february 1`] = `
<nav
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#E0E0E0",
      "display": "flex",
      "justifyContent": "space-evenly",
      "minHeight": "2.5em",
      "paddingLeft": "24px",
      "paddingRight": "24px",
    }
  }
>
  <WithStyles(FormControlLabel)
    control={
      <WithStyles(Radio)
        checked={false}
        onChange={[Function]}
      />
    }
    label="Last 30 days"
  />
  <div>
    <WithStyles(FormControlLabel)
      control={
        <WithStyles(Radio)
          checked={true}
          onChange={[Function]}
        />
      }
      label="Select Month"
    />
    <TextField
      onBlur={[Function]}
      onChange={[Function]}
      required={false}
      select={false}
      type="month"
      value="2018-02"
    />
  </div>
</nav>
`;
